src/solvers.jl

```diff
@@ -38,103 +38,50 @@ function ProblemTraits(m::Model; relaxation=false)
-                alreadydefined = isdefined(Main,pkgname)
-                if !alreadydefined
-                    try
-                        eval(Expr(:import,pkgname))
-                        # if we got here, package works but wasn't loaded,
-                        # print warning.
-                        Base.warn_once(string("The default ", $solvertype, " package is installed but not loaded. In Julia 0.6 and later, an explicit \"using ", pkgname, "\" statement will be required in order for the solver to be detected and used as a default."))
-                    catch
-                        continue
-                    end
-                end
-                return getfield(getfield(Main,pkgname),solvername)()
-            end
-            suggestions = join(["\"$(pkgname)\", " for (pkgname,solvername) in $solvers], ' ')
-            error("No ",$solvertype, " solver detected. Try installing one of the following packages: ", suggestions, " and restarting Julia")
-        end
-    end
-end
-
-function default_solver(traits::ProblemTraits)
+const suggestedsolvers = Dict("LP" => [(:Clp,:ClpSolver),
+                                       (:GLPKMathProgInterface,:GLPKSolverLP),
+                                       (:Gurobi,:GurobiSolver),
+                                       (:CPLEX,:CplexSolver),
+                                       (:Mosek,:MosekSolver),
+                                       (:Xpress,:XpressSolver)],
+                              "MIP" => [(:Cbc,:CbcSolver),
+                                        (:GLPKMathProgInterface,:GLPKSolverMIP),
+                                        (:Gurobi,:GurobiSolver),
+                                        (:CPLEX,:CplexSolver),
+                                        (:Mosek,:MosekSolver),
+                                        (:Xpress,:XpressSolver)],
+                              "QP" => [(:Gurobi,:GurobiSolver),
+                                       (:CPLEX,:CplexSolver),
+                                       (:Mosek,:MosekSolver),
+                                       (:Ipopt,:IpoptSolver),
+                                       (:Xpress,:XpressSolver)],
+                              "SDP" => [(:Mosek,:MosekSolver),
+                                       (:SCS,:SCSSolver)],
+                              "NLP" => [(:Ipopt,:IpoptSolver),
+                                        (:KNITRO,:KnitroSolver),
+                                        (:Mosek,:MosekSolver)],
+                              "Conic" => [(:ECOS,:ECOSSolver),
+                                         (:SCS,:SCSSolver),
+                                         (:Mosek,:MosekSolver)])
+
+function no_solver_error(traits::ProblemTraits)
+
+    # This is pretty coarse and misses out on MIConic, MIQP, etc.
     if traits.nlp
-        defaultNLPsolver()
+        class = "NLP"
     elseif traits.int || traits.sos
-        defaultMIPsolver()
+        class = "MIP"
```

Comment:
Since the issues with the "coarseness" of this function seem to come up with discrete problems, might be good to split this logic out of this loop. Could add a helper function like 
```jl
function supports_MIP(x::Tuple{Symbol,Symbol})
    x[1] in (:Cbc, :Gurobi, :CPLEX, :Knitro, ...)
end
```